smolyay/surrogate.py

```diff
@@ -117,430 +72,694 @@ def data(self):
-                self._data)
-        elif linear_solver == 'inv':
-            self._coefficients = numpy.dot(numpy.linalg.inv(
-                basis_matrix),
-                self._data)
-        elif linear_solver == 'lstsq':
-            self._coefficients = numpy.linalg.lstsq(basis_matrix,
-                                                    self._data,
-                                                    rcond=None)[0]
+        if isinstance(
+            X,
+            (
+                smolyay.samples.UnidimensionalPointSet,
+                smolyay.samples.MultidimensionalPointSet,
+            ),
+        ):
+            X = X.points
+        X, y = self._validate_data(
+            X,
+            y,
+            multi_output=True,
+            y_numeric=True,
+            ensure_2d=True,
+            dtype="numeric",
+        )
+        self._points = X
+        self._data = y
+        oob = any(
+            numpy.any(X[:, i] < self.domain[i][0])
+            or numpy.any(X[:, i] > self.domain[i][1])
+            for i in range(self.num_dimensions)
+        )
+        if oob:
+            raise ValueError("X must lie in domain of surrogate")
+        # create basis matrix
+        num_basis_max = numpy.max([len(p) for p in self._basis_sets])
+        if any(bs.is_complex for bs in self.basis_sets):
+            lookup_table = numpy.zeros(
+                (self.num_dimensions, num_basis_max, len(X)), dtype="complex_"
+            )
+            basis_matrix = numpy.zeros(
+                (len(X), len(self.index_combinations)), dtype="complex_"
+            )
         else:
-            raise ValueError('Solver not recognized')
-
-    @staticmethod
-    def _mapdomain(x, old, new):
-        """Transform the points into new domain.
+            lookup_table = numpy.zeros((self.num_dimensions, num_basis_max, len(X)))
+            basis_matrix = numpy.zeros((len(X), len(self.index_combinations)))
+        # solve for the inputs at all the basis functions
+        for dim in range(self.num_dimensions):
+            for i, basis_fun in enumerate(self.basis_sets[dim]):
+                new_X = basis_fun.domain[0] + (
+                    basis_fun.domain[1] - basis_fun.domain[0]
+                ) * (
+                    (X[:, dim] - self.domain[dim, 0])
+                    / (self.domain[dim, 1] - self.domain[dim, 0])
+                )
+                numpy.clip(new_X, basis_fun.domain[0], basis_fun.domain[1], out=new_X)
+                lookup_table[dim, i, :] = basis_fun(new_X)
+
+        # use lookup table to solve for each term
+        for term, ic in enumerate(self.index_combinations):
+            basis_matrix[:, term] = numpy.prod(
+                [lookup_table[dim, ic[dim], :] for dim in range(len(ic))], axis=0
+            )
+        # solve for coefficients
+        if basis_matrix.shape[0] == basis_matrix.shape[1]:
+            self._coefficients = numpy.linalg.solve(basis_matrix, self._data)
```

Comment:
This may need to be done with a try-except structure. What happens if the matrix is the right shape but under/over determined?